### PR TITLE
Clean and refactor tinkerbell client

### LINKED ISSUE #953: Cleanup the baremetal cluster provider
**User Story**
As a user for a bare-metal cluster, I would like to have a generic design where bare-metal cluster provisioning is not depending on a vendor nor specific infrastructure

**Acceptance Criteria**
- Cleanup any implementation that's bounded to any pre-existing infrastructure 
- Abstract dependencies, especially the metadata server which is used for fetching the bare-metal machine specs, such as IP, MAC Addresses.

### PR DESCRIPTION
**Pull request by moadqassem**
**What this PR does / why we need it**:
The current support for tinkerbell was bounded to a specific infrastructure for testing purposes to validate the functionality of TinkerBell and Machine-Controller. This PR clean and abstract all the needed details to introduce a more of generic driver, 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #953

**Special notes for your reviewer**:
Tests will follow in a separate PR.
**Optional Release Note**:
```release-note
None
```

### CODE REVIEW

**Comment by moadqassem** [comment]
Seems that vSphere setup is broken as this change has nothing to do with the vSphere cloud provider.

File: test/e2e/provisioning/testdata/machinedeployment-baremetal-tinkerbell.yaml [hunk 1]
```diff
@@ -25,6 +25,14 @@ spec:
           cloudProvider: "baremetal"
           cloudProviderSpec:
             driver: "tinkerbell"
+            metadataClient:
+              config:
+                endpoint: ""
+                authConfig:
+                  authMethod: ""
+                  username: ""
```
Comment (mfranczy) [inline test/e2e/provisioning/testdata/machinedeployment-baremetal-tinkerbell.yaml:33]: I would rather use a secret for client credentials

**Review by mfranczy — CHANGES_REQUESTED** [review]
Overall looks good to me, just one small thing about credentials.

File: test/e2e/provisioning/testdata/machinedeployment-baremetal-tinkerbell.yaml (hunk 1 above)
Comment (moadqassem) [reply test/e2e/provisioning/testdata/machinedeployment-baremetal-tinkerbell.yaml]: So done in this commit https://github.com/kubermatic/machine-controller/pull/956/commits/c7e104ea8259550435cfc9a43215ceb1587863cf

**Review by mfranczy — APPROVED** [review]

**Pull request merged**